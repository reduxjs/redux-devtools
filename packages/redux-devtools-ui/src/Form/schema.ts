import { JSONSchema7 } from 'json-schema';

export const schema: JSONSchema7 = {
  title: 'Example form',
  description: 'A simple form example.',
  type: 'object',
  required: ['name'],
  properties: {
    name: {
      type: 'string',
      title: 'Full name',
    },
    age: {
      type: 'integer',
      title: 'Age',
    },
    bio: {
      type: 'string',
      title: 'Bio',
    },
    password: {
      type: 'string',
      title: 'Password',
      minLength: 3,
    },
    multipleChoicesList: {
      type: 'array',
      title: 'A multiple choices list',
      items: {
        type: 'string',
        enum: ['foo', 'bar', 'fuzz'],
      },
      uniqueItems: true,
    },
    numberEnum: {
      type: 'number',
      title: 'Number enum',
      enum: [1, 2, 3],
    },
    numberEnumRadio: {
      type: 'number',
      title: 'Number enum',
      enum: [1, 2, 3],
    },
    integerRange: {
      title: 'Integer range',
      type: 'integer',
      minimum: 42,
      maximum: 100,
    },
  },
};

export const uiSchema = {
  name: {
    'ui:autofocus': true,
  },
  age: {
    'ui:widget': 'updown',
  },
  bio: {
    'ui:widget': 'textarea',
  },
  password: {
    'ui:widget': 'password',
    'ui:help': 'Hint: Make it strong!',
  },
  date: {
    'ui:widget': 'alt-datetime',
  },
  multipleChoicesList: {
    'ui:widget': 'checkboxes',
  },
  numberEnumRadio: {
    'ui:widget': 'radio',
    'ui:options': {
      inline: true,
    },
  },
  integerRange: {
    'ui:widget': 'range',
  },
};

export const formData = {
  name: 'Chuck Norris',
  age: 75,
  bio: 'Roundhouse kicking asses since 1940',
  password: 'noneed',
  integerRange: 52,
};
