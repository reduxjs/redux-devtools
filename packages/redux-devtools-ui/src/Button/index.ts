export { default } from './Button.js';
