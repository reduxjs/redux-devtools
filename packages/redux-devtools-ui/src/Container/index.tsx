import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { useTheme, ThemeData } from '../utils/theme';
import { MainContainerWrapper, ContainerWrapper } from './styles';
import { Theme } from '../themes/default';

interface ContainerFromThemeDataProps {
  children?: React.ReactNode;
  themeData: ThemeData;
  className?: string;
}

const ContainerFromThemeData: React.FunctionComponent<
  ContainerFromThemeDataProps
> = ({ themeData, className, children }) => {
  const theme = useTheme(themeData);
  return (
    <ThemeProvider theme={theme}>
      <MainContainerWrapper className={className}>
        {children}
      </MainContainerWrapper>
    </ThemeProvider>
  );
};

interface Props {
  children?: React.ReactNode;
  themeData?: ThemeData;
  theme?: Theme;
  className?: string;
}

const Container: React.FunctionComponent<Props> = ({
  themeData,
  className,
  theme,
  children,
}) => {
  if (!themeData) {
    return (
      <ContainerWrapper className={className} theme={theme}>
        {children}
      </ContainerWrapper>
    );
  }

  return (
    <ContainerFromThemeData themeData={themeData} className={className}>
      {children}
    </ContainerFromThemeData>
  );
};

Container.propTypes = {
  children: PropTypes.node,
  themeData: PropTypes.any,
  theme: PropTypes.any,
  className: PropTypes.string,
};

export default Container;
