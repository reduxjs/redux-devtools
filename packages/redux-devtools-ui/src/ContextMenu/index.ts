export { default } from './ContextMenu.js';
