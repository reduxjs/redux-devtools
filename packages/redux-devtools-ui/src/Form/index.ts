export { default } from './Form.js';
