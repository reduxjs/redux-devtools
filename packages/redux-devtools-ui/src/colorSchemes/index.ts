export { default } from './default';
export { default as github } from './github';
export { default as atomOneDark } from './atom-one-dark';
export { default as dracula } from './dracula';
export { default as iRBlack } from './ir-black';
export { default as macintosh } from './macintosh';
export { default as materia } from './materia';
export { default as oceanicNext } from './oceanic-next';
export { default as phD } from './phd';
export { default as pico } from './pico';
export { default as solarFlare } from './solar-flare';
export { default as spacemacs } from './spacemacs';
export { default as unikitty } from './unikitty';
export { default as woodland } from './woodland';
