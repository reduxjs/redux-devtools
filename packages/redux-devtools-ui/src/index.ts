export { default as Container } from './Container';
export { default as Button } from './Button';
export { default as ContextMenu } from './ContextMenu';
export { default as Dialog } from './Dialog';
export { default as Editor } from './Editor';
export { default as Form } from './Form';
export { default as Select } from './Select';
export { default as Slider } from './Slider';
export { default as Tabs, Tab } from './Tabs';
export { default as SegmentedControl } from './SegmentedControl';
export { default as Notification } from './Notification';
export * from './Toolbar';

import color from './utils/color';
export const effects = { color };
export { default as createStyledComponent } from './utils/createStyledComponent';
export { Theme, ThemeFromProvider, Scheme } from './utils/theme';
