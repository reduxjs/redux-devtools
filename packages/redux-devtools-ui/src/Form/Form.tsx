import React, { PureComponent, Component } from 'react';
import PropTypes from 'prop-types';
import JSONSchemaForm, { FormProps } from '@rjsf/core';
import { Base16Theme } from 'base16';
import createStyledComponent from '../utils/createStyledComponent';
import styles from './styles';
import Button from '../Button';
import customWidgets from './widgets';

const FormContainer = createStyledComponent(styles, JSONSchemaForm);

export interface Props<T> extends FormProps<T> {
  children?: React.ReactNode;
  submitText?: string;
  primaryButton?: boolean;
  noSubmit?: boolean;
  theme?: Base16Theme;
}

/**
 * Wrapper around [`react-jsonschema-form`](https://github.com/rjsf-team/react-jsonschema-form) with custom widgets.
 */
export default class Form<T> extends (PureComponent || Component)<Props<T>> {
  render() {
    const { widgets, children, submitText, primaryButton, noSubmit, ...rest } =
      this.props;
    return (
      <FormContainer
        {...(rest as Props<unknown>)}
        widgets={{ ...customWidgets, ...widgets }}
      >
        {noSubmit ? (
          <noscript />
        ) : (
          children || (
            <Button
              size="big"
              primary={primaryButton}
              theme={rest.theme}
              type="submit"
            >
              {submitText || 'Submit'}
            </Button>
          )
        )}
      </FormContainer>
    );
  }

  static propTypes = {
    children: PropTypes.any,
    submitText: PropTypes.string,
    primaryButton: PropTypes.bool,
    noSubmit: PropTypes.bool,
    schema: PropTypes.object.isRequired,
    uiSchema: PropTypes.object,
    formData: PropTypes.any,
    widgets: PropTypes.objectOf(
      PropTypes.oneOfType([PropTypes.func, PropTypes.object])
    ),
  };
}
