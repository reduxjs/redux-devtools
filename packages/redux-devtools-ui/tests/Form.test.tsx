import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Form } from '../src';
import { schema, uiSchema, formData } from '../src/Form/schema';

describe('Form', function () {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Form formData={formData} schema={schema} uiSchema={uiSchema} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders with primary button', () => {
    const wrapper = shallow(
      <Form
        primaryButton
        submitText="Custom button"
        formData={formData}
        schema={schema}
        uiSchema={uiSchema}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders with no button', () => {
    const wrapper = shallow(
      <Form formData={formData} schema={schema} uiSchema={uiSchema} noSubmit />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should handle the submit event', () => {
    const onSubmit = jest.fn();
    const wrapper = mount(
      <Form
        formData={formData}
        schema={schema}
        uiSchema={uiSchema}
        onSubmit={onSubmit}
      />
    );

    wrapper.find('form').simulate('submit');
    expect(onSubmit).toBeCalled();
  });
});
