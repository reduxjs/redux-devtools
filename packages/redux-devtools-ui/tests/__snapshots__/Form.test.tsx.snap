// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<Styled(Form)
  formData={
    Object {
      "age": 75,
      "bio": "Roundhouse kicking asses since 1940",
      "integerRange": 52,
      "name": "Chuck Norris",
      "password": "noneed",
    }
  }
  schema={
    Object {
      "description": "A simple form example.",
      "properties": Object {
        "age": Object {
          "title": "Age",
          "type": "integer",
        },
        "bio": Object {
          "title": "Bio",
          "type": "string",
        },
        "integerRange": Object {
          "maximum": 100,
          "minimum": 42,
          "title": "Integer range",
          "type": "integer",
        },
        "multipleChoicesList": Object {
          "items": Object {
            "enum": Array [
              "foo",
              "bar",
              "fuzz",
            ],
            "type": "string",
          },
          "title": "A multiple choices list",
          "type": "array",
          "uniqueItems": true,
        },
        "name": Object {
          "title": "Full name",
          "type": "string",
        },
        "numberEnum": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "title": "Number enum",
          "type": "number",
        },
        "numberEnumRadio": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "title": "Number enum",
          "type": "number",
        },
        "password": Object {
          "minLength": 3,
          "title": "Password",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "title": "Example form",
      "type": "object",
    }
  }
  uiSchema={
    Object {
      "age": Object {
        "ui:widget": "updown",
      },
      "bio": Object {
        "ui:widget": "textarea",
      },
      "date": Object {
        "ui:widget": "alt-datetime",
      },
      "integerRange": Object {
        "ui:widget": "range",
      },
      "multipleChoicesList": Object {
        "ui:widget": "checkboxes",
      },
      "name": Object {
        "ui:autofocus": true,
      },
      "numberEnumRadio": Object {
        "ui:options": Object {
          "inline": true,
        },
        "ui:widget": "radio",
      },
      "password": Object {
        "ui:help": "Hint: Make it strong!",
        "ui:widget": "password",
      },
    }
  }
  widgets={
    Object {
      "RangeWidget": [Function],
      "SelectWidget": [Function],
    }
  }
>
  <Button
    size="big"
    tooltipPosition="top"
    type="submit"
  >
    Submit
  </Button>
</Styled(Form)>
`;

exports[`Form renders with no button 1`] = `
<Styled(Form)
  formData={
    Object {
      "age": 75,
      "bio": "Roundhouse kicking asses since 1940",
      "integerRange": 52,
      "name": "Chuck Norris",
      "password": "noneed",
    }
  }
  schema={
    Object {
      "description": "A simple form example.",
      "properties": Object {
        "age": Object {
          "title": "Age",
          "type": "integer",
        },
        "bio": Object {
          "title": "Bio",
          "type": "string",
        },
        "integerRange": Object {
          "maximum": 100,
          "minimum": 42,
          "title": "Integer range",
          "type": "integer",
        },
        "multipleChoicesList": Object {
          "items": Object {
            "enum": Array [
              "foo",
              "bar",
              "fuzz",
            ],
            "type": "string",
          },
          "title": "A multiple choices list",
          "type": "array",
          "uniqueItems": true,
        },
        "name": Object {
          "title": "Full name",
          "type": "string",
        },
        "numberEnum": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "title": "Number enum",
          "type": "number",
        },
        "numberEnumRadio": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "title": "Number enum",
          "type": "number",
        },
        "password": Object {
          "minLength": 3,
          "title": "Password",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "title": "Example form",
      "type": "object",
    }
  }
  uiSchema={
    Object {
      "age": Object {
        "ui:widget": "updown",
      },
      "bio": Object {
        "ui:widget": "textarea",
      },
      "date": Object {
        "ui:widget": "alt-datetime",
      },
      "integerRange": Object {
        "ui:widget": "range",
      },
      "multipleChoicesList": Object {
        "ui:widget": "checkboxes",
      },
      "name": Object {
        "ui:autofocus": true,
      },
      "numberEnumRadio": Object {
        "ui:options": Object {
          "inline": true,
        },
        "ui:widget": "radio",
      },
      "password": Object {
        "ui:help": "Hint: Make it strong!",
        "ui:widget": "password",
      },
    }
  }
  widgets={
    Object {
      "RangeWidget": [Function],
      "SelectWidget": [Function],
    }
  }
>
  <noscript />
</Styled(Form)>
`;

exports[`Form renders with primary button 1`] = `
<Styled(Form)
  formData={
    Object {
      "age": 75,
      "bio": "Roundhouse kicking asses since 1940",
      "integerRange": 52,
      "name": "Chuck Norris",
      "password": "noneed",
    }
  }
  schema={
    Object {
      "description": "A simple form example.",
      "properties": Object {
        "age": Object {
          "title": "Age",
          "type": "integer",
        },
        "bio": Object {
          "title": "Bio",
          "type": "string",
        },
        "integerRange": Object {
          "maximum": 100,
          "minimum": 42,
          "title": "Integer range",
          "type": "integer",
        },
        "multipleChoicesList": Object {
          "items": Object {
            "enum": Array [
              "foo",
              "bar",
              "fuzz",
            ],
            "type": "string",
          },
          "title": "A multiple choices list",
          "type": "array",
          "uniqueItems": true,
        },
        "name": Object {
          "title": "Full name",
          "type": "string",
        },
        "numberEnum": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "title": "Number enum",
          "type": "number",
        },
        "numberEnumRadio": Object {
          "enum": Array [
            1,
            2,
            3,
          ],
          "title": "Number enum",
          "type": "number",
        },
        "password": Object {
          "minLength": 3,
          "title": "Password",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "title": "Example form",
      "type": "object",
    }
  }
  uiSchema={
    Object {
      "age": Object {
        "ui:widget": "updown",
      },
      "bio": Object {
        "ui:widget": "textarea",
      },
      "date": Object {
        "ui:widget": "alt-datetime",
      },
      "integerRange": Object {
        "ui:widget": "range",
      },
      "multipleChoicesList": Object {
        "ui:widget": "checkboxes",
      },
      "name": Object {
        "ui:autofocus": true,
      },
      "numberEnumRadio": Object {
        "ui:options": Object {
          "inline": true,
        },
        "ui:widget": "radio",
      },
      "password": Object {
        "ui:help": "Hint: Make it strong!",
        "ui:widget": "password",
      },
    }
  }
  widgets={
    Object {
      "RangeWidget": [Function],
      "SelectWidget": [Function],
    }
  }
>
  <Button
    primary={true}
    size="big"
    tooltipPosition="top"
    type="submit"
  >
    Custom button
  </Button>
</Styled(Form)>
`;
